[SPARK-5233][Streaming] Fix error replaying of WAL introduced bug

jerryshao opened the pull request:
Because of lacking of `BlockAllocationEvent` in WAL recovery, the dangled event will mix into the new batch, which will lead to the wrong result. Details can be seen in [SPARK-5233](https://issues.apache.org/jira/browse/SPARK-5233).

jerryshao committed:
Correctly replay the WAL log when recovering from failure

jerryshao committed:
Fix bug in unit test

SparkQA commented:
[Test build #25514 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25514/consoleFull) for   PR 4032 at commit [`4c82be8`](https://github.com/apache/spark/commit/4c82be8e41383a4e4cb6b1e53e47928b3f245f2a).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25514 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25514/consoleFull) for   PR 4032 at commit [`4c82be8`](https://github.com/apache/spark/commit/4c82be8e41383a4e4cb6b1e53e47928b3f245f2a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25514/
Test PASSed.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceivedBlockTracker.scala:
```diff
@@ -106,6 +106,12 @@ private[streaming] class ReceivedBlockTracker(
       timeToAllocatedBlocks(batchTime) = allocatedBlocks
       lastAllocatedBatchTime = batchTime
       allocatedBlocks
+    } else if (batchTime == lastAllocatedBatchTime) {
```
Could you improve the unit test to actually verify the behavior. The unit test line you removed can be modified to verify that calling `batchTime` = `lastAllocatedBatchTime` is a no-op and does no allocation.

jerryshao replied:
OK, I will improve this.

tdas replied:
Update to this comment. See other comment about removing this check altogether. Please update unit test to verify this new behavior.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceivedBlockTracker.scala:
```diff
@@ -106,6 +106,12 @@ private[streaming] class ReceivedBlockTracker(
+      // job need to process again, so the batchTime will be equal to lastAllocatedBatchTime.
+      // This situation will only occurs in recovery time.
+      logWarning(s"Possibly processed batch $batchTime need to be processed again in WAL recovery")
     } else {
```
Actually lets remove this exception completely. Instead any attempts to allocate blocks to a batch such that batch <= lastAllocatedBatchTime, is completely ignored.

jerryshao replied:
IIUC, I think this might not be happened normally, unless the checkpoint operation is so slow. Anyway we should consider this to make the code more robust.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/ReceivedBlockTrackerSuite.scala:
```diff
@@ -82,11 +82,6 @@ class ReceivedBlockTrackerSuite
     receivedBlockTracker.allocateBlocksToBatch(2)
     receivedBlockTracker.getBlocksOfBatchAndStream(2, streamId) shouldBe empty
 
-    // Verify that batch 2 cannot be allocated again
```
Please add test to verify that `allocateBlocksToBatch(x)` where x<=2 does nothing.

tdas commented:
I understand this patch now. Please update it based on my comments, and test it in your harness to make sure that it addresses the exception problem.

jerryshao commented:
OK, will do, thanks a lot for your comments.

jerryshao committed:
Address the comments

SparkQA commented:
[Test build #25964 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25964/consoleFull) for   PR 4032 at commit [`a237c75`](https://github.com/apache/spark/commit/a237c75f87518b26245fd688de9bbae4bc151ae2).
 * This patch merges cleanly.

jerryshao commented:
Hey @tdas , I've updated the code and rebased the branch according to your comments, also with several rounds of test in my local set, the previous exception I reported is gone :).

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala:
```diff
@@ -206,9 +208,13 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
-    timesToReschedule.foreach(time =>
+    timesToReschedule.foreach { time =>
+      // Allocate the related blocks when recovering from failure, because some added but not
+      // allocated block is dangled in the queue after recovering, we have to insert some block
```
Grammar fix
"..some blocks that were added but not allocated, are dangling in the queue after recovering. We have to allocate thoe blocks to the next batch, which is the batch they were supposed to go to."

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/ReceivedBlockTrackerSuite.scala:
```diff
@@ -82,15 +82,13 @@ class ReceivedBlockTrackerSuite
-      receivedBlockTracker.allocateBlocksToBatch(1)
-    }
+    receivedBlockTracker.allocateBlocksToBatch(2)
+    receivedBlockTracker.getBlocksOfBatchAndStream(2, streamId) shouldBe empty
```
Could you also test that even if there are new unallocated blocks, calling `receivedBlockTracker.allocateBlocksToBatch(2)` does not allocate those blocks? I dont think that is covered in this test.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceivedBlockTracker.scala:
```diff
@@ -107,8 +107,14 @@ private[streaming] class ReceivedBlockTracker(
+      // 2. Slow checkpointing makes recovered batch time older than WAL recovered
+      // lastAllocatedBatchTime.
+      // This situation will only occurs in recovery time.
+      logWarning(s"Possibly processed batch $batchTime need to be processed again in WAL recovery")
```
i am not sure we even should throw a warning. Warnings are thrown when something bad may have happened. In this case, if the code reaches here, nothing bad or unexpected has happened.

jerryshao replied:
At least I think we should inform user this batch is processed or half-processed before, so user may pay attention to this, whether to clean or overwrite the previously processed result. I'm downgrading to logInfo, what's your opinion?

tdas commented:
looks almost good. some minor comments.

SparkQA commented:
[Test build #25964 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25964/consoleFull) for   PR 4032 at commit [`a237c75`](https://github.com/apache/spark/commit/a237c75f87518b26245fd688de9bbae4bc151ae2).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25964/
Test PASSed.

jerryshao committed:
Further address the comments

SparkQA commented:
[Test build #25966 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25966/consoleFull) for   PR 4032 at commit [`f0b0c0b`](https://github.com/apache/spark/commit/f0b0c0bcc8db551c57c1590c8a12b01f5f0ae2d0).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25966 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25966/consoleFull) for   PR 4032 at commit [`f0b0c0b`](https://github.com/apache/spark/commit/f0b0c0bcc8db551c57c1590c8a12b01f5f0ae2d0).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25966/
Test PASSed.

tdas commented:
LGTM. Merging this. Thanks!!!

tdas commented:
LGTM. Merging this. Thanks!

jerryshao commented:
Thanks for your reviewing :)
